File: advent2021/day3.ts
import fs = require("fs");
import readline = require("readline");

async function processLineByLine() {
  const fileStream = fs.createReadStream("inputs/day3.txt");

  const rl = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity,
  });

  const n: number = 12;
  let data: number[] = new Array(n).fill(0);
  let cnt: number = 0;
  for await (const line of rl) {
    for (let i = 0; i < line.length; i++) {
      data[i] += Number(line[i]);
    }
    cnt++;
  }

  let gamma: string = "";
  let epsilon: string = "";
  data.forEach((num) => {
    let result: boolean = num > cnt / 2;
    gamma += Number(result);
    epsilon += Number(!result);
  });

  console.log(parseInt(gamma, 2) * parseInt(epsilon, 2));
}

async function processLineByLine2() {
  const fileStream = fs.createReadStream("inputs/day3.txt");

  const rl = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity,
  });

  const n: number = 12;
  let data: number[] = new Array(n).fill(0);
  let cnt: number = 0;
  for await (const line of rl) {
    for (let i = 0; i < line.length; i++) {
      data[i] += Number(line[i]);
    }
    cnt++;
  }

  let gamma: string = "";
  let epsilon: string = "";
  data.forEach((num) => {
    let result: boolean = num > cnt / 2;
    gamma += Number(result);
    epsilon += Number(!result);
  });

  console.log(parseInt(gamma, 2) * parseInt(epsilon, 2));
}

// processLineByLine();
processLineByLine2();

File: advent2021/day3.py
def part1():
    file=open("inputs/day3.txt","rt")
    data=file.readlines()
    n=12
    l=len(data)
    bitCnts=[0]*n
    cnt=0
    for i in range(l):
        for j in range(len(data[i])-1):
            bitCnts[j]+=int(data[i][j])
        cnt+=1
    g=''
    e=''
    for i in range(n):
        r=bitCnts[i]>cnt/2
        g+=str(int(r))
        e+=str(int(not r))
    print(int(g,2)*int(e,2))

def filter_data_by_index(data,idx,mostCommon=True):
    l=len(data)
    if l==1:
        return data[0]
    cnt=0
    for i in range(l):
        cnt+=int(data[i][idx])
    filtered=[]
    if mostCommon:
        for i in range(l):
            if int(data[i][idx])==int(cnt>=l/2):
                filtered.append(data[i])
    else:
        for i in range(l):
            if int(data[i][idx])!=int(cnt>=l/2):
                filtered.append(data[i])
    return filter_data_by_index(filtered,idx+1,mostCommon)

def part2():
    file=open("inputs/day3.txt","rt")
    data=file.readlines()
    print(int(filter_data_by_index(data,0),2)*int(filter_data_by_index(data,0,False),2))

part1()
part2()